components/templates/src/filters.rs

```diff
@@ -80,20 +81,49 @@ pub fn base64_decode<S: BuildHasher>(
     Ok(to_value(as_str).unwrap())
 }
 
-pub fn replace_re(value: &Value, args: &HashMap<String, Value>) -> TeraResult<Value> {
-    let text = try_get_value!("replace_re", "value", String, value);
-    let pattern = match args.get("pattern") {
-        Some(val) => try_get_value!("replace_re", "pattern", String, val),
-        None => return Err(TeraError::msg("Filter `replace_re` expected an arg called `pattern`")),
-    };
-    let rep = match args.get("rep") {
-        Some(val) => try_get_value!("replace_re", "rep", String, val),
-        None => return Err(TeraError::msg("Filter `replace_re` expected an arg called `rep`")),
-    };
-    let pattern_re = Regex::new(&pattern)
-        .map_err(|e| format!("`replace_re`: failed to compile regex: {}", e))?;
-    let replaced = pattern_re.replace_all(&text, &rep);
-    Ok(to_value(replaced).unwrap())
+#[derive(Debug)]
+pub struct ReplaceReFilter {
+    re_cache: Arc<Mutex<HashMap<String, Regex>>>,
+}
+
+impl ReplaceReFilter {
+    pub fn new() -> Self {
+        return Self { re_cache: Arc::new(Mutex::new(HashMap::new())) };
+    }
+}
+
+impl TeraFilter for ReplaceReFilter {
+    fn filter(&self, value: &Value, args: &HashMap<String, Value>) -> TeraResult<Value> {
+        let text = try_get_value!("replace_re", "value", String, value);
+        let pattern = match args.get("pattern") {
+            Some(val) => try_get_value!("replace_re", "pattern", String, val),
+            None => {
+                return Err(TeraError::msg("Filter `replace_re` expected an arg called `pattern`"))
+            }
+        };
+        let rep = match args.get("rep") {
+            Some(val) => try_get_value!("replace_re", "replace", String, val),
+            None => {
+                return Err(TeraError::msg("Filter `replace_re` expected an arg called `replace`"))
+            }
+        };
+
+        let mut cache = self.re_cache.lock().expect("re_cache lock");
+        let pattern_re = {
+            match cache.get(&pattern) {
+                Some(pat) => pat.clone(),
```

Comment:
You don't need to clone it, you can call replace_all here and return replace in L112